packages/api/lib/ems.js

```diff
@@ -0,0 +1,281 @@
+  case 'insertTime':
+    // milliseconds to string
+    result = (metadata) ? moment(new Date(metadata)).format('YYYY-MM-DD hh:mmA') : metadata;
+    break;
+  case 'lastUpdate':
+  case 'processingStartDateTime':
+  case 'processingEndDateTime':
+  case 'beginningDateTime':
+  case 'endingDateTime':
+  case 'productionDateTime':
+    // string to different format string
+    result = (metadata) ? moment(Date.parse(metadata)).format('YYYY-MM-DD hh:mmA') : metadata;
+    break;
+  default:
+    break;
+  }
+  return result;
+}
+
+/**
+ * build EMS records from es granules
+ *
+ * @param {Object} mapping - mapping of EMS fields to granule fields
+ * @param {Object} granules - es granules
+ * @returns {Array<string>} EMS records
+ */
+function buildEMSRecords(mapping, granules) {
+  const records = granules.map((g) => {
+    const record = Object.keys(mapping).map((k) => getEmsFieldFromGranField(g, k, mapping[k]));
+    return record.join('|&|');
+  });
+  return records;
+}
+
+/**
+ * upload a report to s3, a rev file is created if the report already exists in s3
+ *
+ * @param {string} filename - file to be upload to s3
+ * @returns {string} - uploaded file in s3
+ */
+async function uploadReportToS3(filename) {
+  const bucket = process.env.bucket;
+  const originalKey = `${process.env.stackName}/ems/${path.basename(filename)}`;
+  let key = originalKey;
+  let exists = await aws.fileExists(bucket, key);
+  let i = 1;
+  while (exists) {
+    key = `${originalKey}.rev${i}`;
+    exists = await aws.fileExists(bucket, key);
+    i += 1;
+  }
+
+  await aws.s3().putObject({
+    Bucket: bucket,
+    Key: key,
+    Body: fs.createReadStream(filename)
+  }).promise();
+
+  //TODO uncomment
+  //fs.unlinkSync(filename);
+  const s3Uri = `s3://${bucket}/${key}`;
+  log.info(`uploaded ${s3Uri}`);
+  return s3Uri;
+}
+
+/**
+ * build report file name
+ * The report filename is in format:
+ * <YYYYMMDD> _<Provider>_<FileType>_<DataSource>.flt.rev<1-n>
+ *
+ * @param {string} reportType - report type (ingest, archive, delete)
+ * @returns {string} - report file name
+ */
+function buildReportFileName(reportType) {
+  // DataSource: designates the database table name or data source file/table name
+  // use stackname as DataSource for now
+  const provider = process.env.ems_provider || 'cumulus';
+  const dataSource = process.env.stackName;
+  const datestring = moment().format('YYYYMMDD');
+  let fileType = 'Ing';
+  if (reportType === 'archive') fileType = 'Arch';
+  else if (reportType === 'delete') fileType = 'ArchDel';
+  return `${datestring}_${provider}_${fileType}_${dataSource}.flt`;
+}
+
+/**
+ * generate an EMS report
+ *
+ * @param {string} reportType - report type (ingest, archive, delete)
+ * @param {string} startTime - start time of the records
+ * @param {string} endTime - end time of the records
+ * @returns {Object} - report type and its file path {reportType, file}
+ */
+async function generateReport(reportType, startTime, endTime) {
```

Comment:
We might want to throw an error for a reportType outside of ingest, archive, delete

Reply:
it's done.